packages/react/src/hooks/program.ts

```diff
@@ -0,0 +1,200 @@
+      delay = Math.min(delay * 1.2, CONFIRM_TRANSATION_MAX_MILLISECOND_DELAY);
+      retryCount++;
+      if (error instanceof Error && error.message.includes("Transaction failed")) {
+        throw error;
+      }
+    }
+  }
+}
+
+export function createProgramHook<
+  TInstructions extends Record<string, InstructionBuilder<any>>,
+  TAccounts extends Record<string, AccountFetcher<any>>,
+>(config: ProgramHookConfig<TInstructions, TAccounts>) {
+  const defaultCommitment = config.commitment === undefined ? "confirmed" : config.commitment;
+
+  function useProgramMutation<TInstructionName extends keyof TInstructions>(
+    input: UseProgramMutationInput<TInstructions, TInstructionName>,
+  ) {
+    const queryClient = useQueryClient();
+    const { instruction, commitment: mutationCommitment, ...options } = input;
+    const instructionFn = config.instructions[instruction];
+
+    return useMutation({
+      ...options,
+      mutationFn: async (input: {
+        commitment?: CommitmentLevel;
+        params: Parameters<TInstructions[TInstructionName]>[0];
+        rpc: SolanaClient['rpc'];
+        signAndSend: SignAndSendFn;
+        signer: TransactionSendingSigner;
+      }) => {
+        if (!input.params) {
+          throw new Error("Instruction params are required");
+        }
+        const { params, signer, commitment: inputCommitment, rpc, signAndSend } = input;
+        const commitmentToUse = inputCommitment ?? mutationCommitment ?? defaultCommitment;
+
+        const instruction = instructionFn(params);
+
+        const { simulateTransaction } = createSolanaClient({
+          urlOrMoniker: "devnet",
+        });
+
+        const { value: latestBlockhash } = await rpc.getLatestBlockhash().send();
+
+        const transaction = createTransaction({
+          feePayer: signer,
+          instructions: [instruction],
+          latestBlockhash,
+          version: 0,
+        });
+
+        const simulation = await simulateTransaction(transaction);
+        console.log("simulation", simulation);
+
+        const signature = await signAndSend(instruction, signer);
+
+        if (commitmentToUse) {
+          await waitForConfirmation(rpc, signature as Signature, commitmentToUse);
+        }
+
+        await queryClient.refetchQueries({
+          queryKey: [GILL_HOOK_CLIENT_KEY, config.programAddress],
+        });
+
+        return signature;
+      },
+    });
+  }
+
+  function useProgramQuery<TAccountName extends keyof TAccounts>(input: UseProgramQueryInput<TAccounts, TAccountName>) {
+    const { account, address, rpc, ...options } = input;
+    const accountFetcher = config.accounts[account];
+
+    return useQuery({
+      ...options,
+      enabled: options.enabled !== false && !!address && !!rpc,
+      queryFn: async () => {
+        const data = await accountFetcher(rpc, address);
+        return data;
+      },
+      queryKey: [GILL_HOOK_CLIENT_KEY, config.programAddress, account, address],
```

Comment:
I haven't tested it, but wondering if rpc should be included? Would that help invalidate if a user switched networks?